### PR TITLE
feat(HACBS-1552): update upload sbom task

### PR DESCRIPTION
**Pull request by mmalina**
Adapt the task to support multiple images on the input.

Also, add a second step to run the upload_sbom script for each sbom file.

### CODE REVIEW

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (mmalina) [inline catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:99]: I guess we should probably wait for the script to be available before merging this PR? (That work is currently blocked by https://issues.redhat.com/browse/ISV-3314 )

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (mmalina) [inline catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:90]: Maybe we should reconsider the `set -x` that we have everywhere. In general it makes the log maybe too chatty. But also, in this case it will spit out your secrets to the log.

File: catalog/task/push-sbom-to-pyxis/0.1/README.md
Comment (scoheb) [inline catalog/task/push-sbom-to-pyxis/0.1/README.md:9]: these binary image URLs contain the SBOM?

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (scoheb) [reply catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:90]: good point

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (scoheb) [reply catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:99]: I am ok to merge this before the script is available if it helps us shake out any bugs along the way.

File: catalog/task/push-sbom-to-pyxis/0.1/README.md
Comment (johnbieren) [inline catalog/task/push-sbom-to-pyxis/0.1/README.md:10]: nit: list *of* Pyxis image IDs

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (johnbieren) [inline catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:19]: nit: list *of* Pyxis image IDs

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (johnbieren) [reply catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:90]: Yeah there is no `set -x` in the `create-pyxis-image` task

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (mmalina) [reply catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml]: Thanks, fixed.

File: catalog/task/push-sbom-to-pyxis/0.1/README.md
Comment (mmalina) [reply catalog/task/push-sbom-to-pyxis/0.1/README.md]: Thanks, fixed.

File: catalog/task/push-sbom-to-pyxis/0.1/README.md
Comment (mmalina) [reply catalog/task/push-sbom-to-pyxis/0.1/README.md:9]: @scoheb they do, but that's not how we get them. The build pipeline will call `cosign attach sbom` to push a separate image with just the sbom. See here: https://github.com/redhat-appstudio/build-definitions/pull/288/files#diff-3582b1f53880964d98bc28af3a4c933d29c16f79436d21cca84e481e410c675cR271

This is an example from my testing:
```
cosign attach sbom --sbom ~/tmp/sbom/sbom-cyclonedx.json --type cyclonedx quay.io/mmalina/ubi9:latest
WARNING: Attaching SBOMs this way does not sign them. If you want to sign them, use 'cosign attest --predicate /Users/mmalina/tmp/sbom/sbom-cyclonedx.json --key <key path>' or 'cosign sign --key <key path> --attachment sbom <image uri>'.
Uploading SBOM file for [quay.io/mmalina/ubi9:latest] to [quay.io/mmalina/ubi9:sha256-648ae21c84ae3c2e747af5c981674606324b659dcc15c522717ba8d77dcc178e.sbom] with mediaType [application/vnd.cyclonedx+json].
```

You can see the binary image and then the sbom image here: https://quay.io/repository/mmalina/ubi9?tab=tags

So under the hood, cosign will create a new tag for this sbom image. You can customize what tag is used, but both build and us will use the default, so it should just work.

Comment (scoheb) [reply catalog/task/push-sbom-to-pyxis/0.1/README.md:9]: maybe I would add this text to description: `Space separated list of binary image URLs that are used to obtain SBOMs via cosign`?

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (mmalina) [reply catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:90]: OK, I removed the `-x` here.

> Yeah there is no `set -x` in the `create-pyxis-image` task

Ah, you're right, I didn't notice it. But `create-pyxis-image` also doesn't have `set -e` which means that if one command in the script fails, the execution will carry on. E.g. if `skopeo inspect` fails, `create_container_image` will still be called. That doesn't seem right, WDYT @johnbieren ?

File: catalog/task/push-sbom-to-pyxis/0.1/README.md
Comment (mmalina) [reply catalog/task/push-sbom-to-pyxis/0.1/README.md]: Ok, sounds good. Fixed.

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (scoheb) [reply catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:99]: keep in mind we do not have any customers yet onboarded...

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (johnbieren) [reply catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:90]: https://github.com/hacbs-release/release-bundles/pull/93

**Review by johnbieren — APPROVED** [review]

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (davidmogar) [reply catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:99]: Yep, I'm ok with that.

Comment (mmalina) [reply catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml]: Never mind. I didn't manage to merge this before I left and now the `upload_sbom` is under review, so I will wait for that: https://github.com/redhat-appstudio/release-service-utils/pull/33

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (mmalina) [reply catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml]: This is done now. In fact, based on feedback from Pyxis team, I switched my script to only use GraphQL exclusively. So I will not need the REST API URL at all.

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (mmalina) [inline catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:60]: Note to self: This will need to be updated to the right image once https://github.com/redhat-appstudio/release-service-utils/pull/33 is merged and the new image is pushed to Quay. Also, at that point I will test it once again before merging.

**Comment by mmalina** [comment]
@scoheb @johnbieren can you review again, please? I rebased, but also updated the PR slightly based on changes to the script: https://github.com/redhat-appstudio/release-service-utils/pull/33

The change is only in the `push-sbom-files-to-pyxis` task - `PYXIS_GRAPHQL_API` env var is used now instead of `PYXIS_URL`.

File: catalog/task/push-sbom-to-pyxis/0.1/README.md
Comment (scoheb) [inline catalog/task/push-sbom-to-pyxis/0.1/README.md:11]: maybe mention the `key` and `cert` required keys?

File: catalog/pipeline/push-to-external-registry/0.6/README.md
Comment (johnbieren) [inline catalog/pipeline/push-to-external-registry/0.6/README.md:19]: I don't think this should end in a `.` as no other parameters do (same for everywhere it is listed)

File: catalog/task/push-sbom-to-pyxis/0.1/README.md
Comment (johnbieren) [inline catalog/task/push-sbom-to-pyxis/0.1/README.md:26]: The example usage is no longer correct now that we use bundle resolvers. Can you either fix it or drop it from the README (seems other tasks don't have one, although it is quite nice)

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (johnbieren) [inline catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:34]: I think when you change the other image line you should change this one too. I know it isn't required, but it would be confusing to have two different image tasks of the same image in one task

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (theflockers) [inline catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:84]: Instead of using $param.server point to stage or prod, would not be better have the URL itself as a parameter?

**Review by theflockers — COMMENTED** [review]
Some notes.

File: catalog/task/create-pyxis-image/0.3/create-pyxis-image.yaml
Comment (mmalina) [reply catalog/task/create-pyxis-image/0.3/create-pyxis-image.yaml:21]: I had to add the quotes, otherwise it wouldn't be a valid yaml because of the `:` in the string.

File: catalog/pipeline/push-to-external-registry/0.6/push-to-external-registry.yaml
Comment (mmalina) [reply catalog/pipeline/push-to-external-registry/0.6/push-to-external-registry.yaml:53]: I had to add the quotes, otherwise it wouldn't be a valid yaml because of the `:` in the string.

File: catalog/pipeline/push-to-external-registry/0.6/README.md
Comment (mmalina) [reply catalog/pipeline/push-to-external-registry/0.6/README.md]: It felt natural to use full sentences once it's over one sentence long. But you're right, other strings miss the final `.` as well. Fixed.

File: catalog/task/push-sbom-to-pyxis/0.1/README.md
Comment (mmalina) [reply catalog/task/push-sbom-to-pyxis/0.1/README.md]: Thanks, removed. The example section definitely wasn't my invention, not sure where I copied it from. But it's kind of a duplicate, since the same thing is in the sample yaml.

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (mmalina) [reply catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:34]: OK, I'll do that.

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (mmalina) [reply catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:84]: Maybe, but I'm using the same approach as `catalog/task/create-pyxis-image/0.3/create-pyxis-image.yaml` here - I think it's good to stay consistent.

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (theflockers) [reply catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:12]: Yep. I had the same conversation some weeks ago because I wanted also to do small changes on something not in  use yet but then the conclusion was that we should always create a new version.

Comment (theflockers) [reply catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:12]: Just rephrasing what I said. The conclusion was in the conversation I had (maybe with Scott) - Indeed, this was not a team consensus.

File: catalog/task/create-pyxis-image/0.3/create-pyxis-image.yaml
Comment (mmalina) [reply catalog/task/create-pyxis-image/0.3/create-pyxis-image.yaml]: I agree it's nicer, thanks. Fixed.

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (mmalina) [reply catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:12]: Hmm, good point. But I vaguely remember a similar conversation where the conclusion was that if it wasn't used yet, it's not needed. And it would make sense - imho the reason for new versions is to avoid breaking some existing workflow. But maybe my memory is failing me :) In my opinion adding a new version here would just add more clutter with no benefit. I guess what matters is if it's ok with you as reviewers :)

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (theflockers) [reply catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:85]: I understand the point although I think that in case (for any reason) the URL changes, or in case of a big downtime in pyxis that would require a change to a temp url, , we would need to release a new version of the script. I would prefer just change the parameter in the strategy.

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (theflockers) [reply catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:12]: I think in cases like this one, this would only help to create/enforce the habbit as nothing is yet using it.  Anyway for the time being, all good changing the same ver.

File: catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml
Comment (mmalina) [reply catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:85]: That's a good point. But on the other hand I don't think it's likely that urls such as Pyxis endpoints will change often - Pyxis is used by many things and it cannot just do that without some big announcements and time for users to adapt etc. So I don't think it's a big concern.

@davidmogar wdyt of this?

Comment (theflockers) [reply catalog/task/push-sbom-to-pyxis/0.1/push-sbom-to-pyxis.yaml:85]: I am ok going this way if all agree. For the record, I am a pessimist - or maybe my former sysadmin career is talking louder - that what is unlikely to happen, will happen sooner or later. The most important is being aware. If it happens we at least already know how to fix it.

**Review by johnbieren — COMMENTED** [review]
Not marking it as approved so a bot or someone doesn't accidentally merge it before the image line is updated, but the PR lgtm

**Comment by mmalina** [comment]
> Not marking it as approved so a bot or someone doesn't accidentally merge it before the image line is updated, but the PR lgtm

@johnbieren thanks. But FTR, that's what the open threads are for ;) They will prevent merging.

> Unresolved conversations
> 3 conversations must be resolved before merging.

**Review by kasemAlem — DISMISSED** [review]

**Comment by mmalina** [comment]
> Fair enough, no idea what rules this bot follows though. It certainly doesn't listen to me

Which bot? I don't think there is a bot in this repo. Only the release-service repo has the openshift CI bots set up. There is no automatic merging in this repo.

**Comment by johnbieren** [comment]
> > Fair enough, no idea what rules this bot follows though. It certainly doesn't listen to me
> 
> Which bot? I don't think there is a bot in this repo. Only the release-service repo has the openshift CI bots set up. There is no automatic merging in this repo.

That's the one I meant. I wasn't involved in adding it to the release-service, so I wasn't sure what repos it was added to

**Review by johnbieren — DISMISSED** [review]

**Review by kasemAlem — DISMISSED** [review]

**Review by happybhati — DISMISSED** [review]

**Comment by mmalina** [comment]
While testing yesterday, I found some bugs, mostly in the Python script: https://github.com/redhat-appstudio/release-service-utils/pull/36 but also in the task, so I added a fix in a [new commit](https://github.com/redhat-appstudio/release-service-bundles/pull/92/commits/83db1e95dc68299be2e0c10f8f02a3d9c8776072) in this PR.

Note that I'll still have to update the image references once the release-service-utils PR is merged, so I will ask for the final review and approval when that happens.

**Review by johnbieren — DISMISSED** [review]

**Comment by mmalina** [comment]
I updated the image reference to use the new image that was built after the last bugfix: https://github.com/redhat-appstudio/release-service-utils/pull/36

@johnbieren if you could kindly approve one more time, I really hope this is the last time :) Much appreciated. My goal is to do one final test tomorrow morning and then merge.

**Review by johnbieren — DISMISSED** [review]

**Review by davidmogar — APPROVED** [review]

**Review by theflockers — APPROVED** [review]

**Pull request merged**
